config/nrfconnect/chip-module/CMakeLists.txt

```diff
@@ -305,6 +306,19 @@ if (CONFIG_CHIP_LIB_SHELL)
     target_link_options(chip INTERFACE -Wl,--whole-archive -lCHIPShell -Wl,--no-whole-archive)
 endif()
 
+if (CONFIG_CHIP_MALLOC_SYS_HEAP)
+    target_link_options(chip INTERFACE
+        -Wl,--wrap=malloc
+        -Wl,--wrap=calloc
+        -Wl,--wrap=realloc
+        -Wl,--wrap=free
+        -Wl,--wrap=_malloc_r
+        -Wl,--wrap=_calloc_r
+        -Wl,--wrap=_realloc_r
+        -Wl,--wrap=_free_r
```

Comment:
Making the wrapping decision at this point may clash with downstream users who may want to wrap those for their own reasons. Consider instead providing some API surface to get the heap used, and guidance how to obtain/override the information, without needs to force a given heap implementation.

I will approve this PR, since it is done to paltform code only, and doesn't impact me directly, but based on prior experience, I have found that deep-seated linker wrapping choices made by anyone other than a particular product's final developer can lead to surprises.

Reply:
Good point. Added a separate switch `CHIP_MALLOC_SYS_HEAP_OVERRIDE` for wrapping the default symbols. `CHIP_MALLOC_SYS_HEAP` alone simply enables custom memory allocation functions, but doesn't replace the defaults.